src/toxcore/dht/server/mod.rs

```diff
@@ -746,6 +748,24 @@ impl Server {
         }
     }
 
+    /// Update returned socket address and time of receiving packet
+    fn update_returned_sock(&self, node: &PackedNode, close_nodes: &mut Kbucket, friends: &mut Vec<DhtFriend>) {
+        if self.pk == node.pk {
+            if let Some(node_to_update) = close_nodes.get_node_mut(&node.pk) {
+                node_to_update.update_returned_sock(node.saddr);
+            } else {}
+        }
+
+        friends.iter_mut()
+            .for_each(|friend| {
+                if friend.pk == node.pk {
+                    if let Some(node_to_update) = friend.close_nodes.get_node_mut(&friend.pk, &node.pk) {
```

Comment:
Wrong pk is used. You should update node that sent `NodesResponse`. `node.pk` should be friend's pk.